File: encoding/bench/bench.py
import time

REPS = 1 << 18
TEXT = "Hello \u00e9 \u03a9 \u20ac \u4e2d \U0001f600\n" * REPS
OCTETS = TEXT.encode()
HEX = OCTETS.hex()


def chk(xs) -> int:
    h, n = 0, 0
    for x in xs:
        h = (h * 31 + x) & 0xFFFFFFFF
        n += 1
    return (h + n) & 0xFFFFFFFF


def lap(name, f, x, codes):
    t0 = time.perf_counter()
    y = f(x)
    ms = (time.perf_counter() - t0) * 1000
    print(f"{name}\t{ms:.3f}\t{chk(codes(y))}")


lap("utf8_encode", str.encode, TEXT, iter)
lap("utf8_decode", lambda b: b.decode("utf-8", "replace"), OCTETS, lambda s: map(ord, s))
lap("hex_encode", bytes.hex, OCTETS, lambda s: map(ord, s))
lap("hex_decode", bytes.fromhex, HEX, iter)

File: encoding/bench/bench.ts
// Run with node or bun. Buffer is the Node standard library's hex codec.
const REPS = 1 << 18;
const TEXT = "Hello \u00e9 \u03a9 \u20ac \u4e2d \u{1f600}\n".repeat(REPS);
const OCTETS = new TextEncoder().encode(TEXT);
const HEX = Buffer.from(OCTETS).toString("hex");

function chk(xs: Iterable<number>): number {
  let h = 0, n = 0;
  for (const x of xs) {
    h = (Math.imul(h, 31) + x) >>> 0;
    n++;
  }
  return (h + n) >>> 0;
}

function* codes(s: string): Iterable<number> {
  for (const c of s) yield c.codePointAt(0)!;
}

function lap<A, B>(name: string, f: (a: A) => B, x: A, toCodes: (b: B) => Iterable<number>) {
  const t0 = performance.now();
  const y = f(x);
  const ms = performance.now() - t0;
  console.log(`${name}\t${ms.toFixed(3)}\t${chk(toCodes(y))}`);
}

const enc = new TextEncoder();
const dec = new TextDecoder();
lap("utf8_encode", (s: string) => enc.encode(s), TEXT, (b) => b);
lap("utf8_decode", (b: Uint8Array) => dec.decode(b), OCTETS, codes);
lap("hex_encode", (b: Uint8Array) => Buffer.from(b.buffer, b.byteOffset, b.length).toString("hex"), OCTETS, codes);
lap("hex_decode", (s: string) => Buffer.from(s, "hex"), HEX, (b) => b);
